src/test_ambit.cc

```diff
@@ -0,0 +1,1999 @@
+        ::std::make_tuple(
+            kPass, test_chain_multiply3_batched,
+            "D4(\"ijkl\") += batched(\"j\",A4(\"ijmn\") * B2(\"km\") * C2(\"ln\"))"),
+        ::std::make_tuple(
+            kPass, test_chain_multiply4_batched,
+            "D4(\"ijkl\") -= batched(\"kl\",A4(\"ijmn\") * B2(\"km\") * C2(\"ln\"))"),
+        ::std::make_tuple(
+            kPass, test_batched,
+            "D4(\"ijkl\") += batched(\"ijkl\", A4(\"ijmn\") * B2(\"km\") * B2(\"ln\"))"),
+        ::std::make_tuple(
+            kPass, test_batched_with_factor,
+            "C2(\"ijrs\") = batched(\"r\", 0.5 * A(\"abrs\") * B(\"ijab\"))"),
+        ::std::make_tuple(
+            kPass, test_batched_with_factor_permute,
+            "C2(\"ijrs\") = batched(\"r\", 0.5 * A(\"abrs\") * B(\"ijba\"))"),
+    };
+
+    ::std::vector<::std::tuple<::std::string, TestResult, double>> results;
+
+    outfile->Printf("\n\n==> TEST AMBIT <==\n");
+
+    outfile->Printf("\n %-50s %12s %s", "Description", "Max. error", "Result");
+    outfile->Printf("\n %s", ::std::string(83, '-').c_str());
+
+    double max_error = 0.0;
+
+    bool success = true;
+    for (auto test_function : test_functions)
+    {
+        outfile->Printf("\n %-60s", ::std::get<2>(test_function));
+        double result = 0.0;
+        TestResult tresult = kPass, report_result = kPass;
+        ::std::string exception;
+        try
+        {
+            result = ::std::get<1>(test_function)();
+            max_error = ::std::max(result, max_error);
+
+            // Did the test pass based on returned value?
+            tresult = ::std::fabs(result) < zero ? kPass : kFail;
+            // Was the tresult the expected result? If so color green else red.
+            report_result =
+                tresult == ::std::get<0>(test_function) ? kPass : kFail;
+        }
+        catch (::std::exception &e)
+        {
+            // was an exception expected?
+            tresult = kException;
+            report_result =
+                tresult == ::std::get<0>(test_function) ? kPass : kException;
+
+            if (report_result == kException)
+            {
+                max_error = ::std::numeric_limits<double>::max();
+                exception = e.what();
+            }
+        }
+        outfile->Printf(" %7e", result);
+        switch (tresult)
+        {
+        case kPass:
+            outfile->Printf(" Passed");
+            break;
+        case kFail:
+            outfile->Printf(" Failed");
+            break;
+        default:
+            outfile->Printf(" Exception");
+        }
+
+        if (report_result == kException)
+            outfile->Printf("\n    Unexpected: %s", exception.c_str());
+        if (report_result != kPass)
+            success = false;
+    }
+    outfile->Printf("\n %s", ::std::string(83, '-').c_str());
+    outfile->Printf("\n Tests: %s\n", success ? "All passed" : "Some failed");
+
```

Comment:
Is this file basically copied from ambit?

Reply:
I think so.